protocols/abci/addressing-1174.md

```diff
@@ -0,0 +1,230 @@
+is not able to fulfill coherence and determinism requirements,
+CometBFT cannot guarantee consensus _termination_ in all runs of the system.
+As a result, the application designers must start considering both CometBFT and their application
+as one monolithic block, in order to reason about termination.
+We thus lose the modularity provided when fulfilling the ABCI 1.0 requirements.
+Remember that CometBFT's consensus algorithm (Tendermint) is a well-known algorithm that
+has been studied, reviewed, formally analyzed, model-checked, etc.
+The combination of CometBFT and an arbitrary application as one single algorithm cannot
+leverage that extensive body of research applied to the Tendermint algorithm.
+This situation is risky and undesirable.
+
+So, the questions that arises are the following.
+Can we come up with a set of weaker requirements
+that applications unable to fulfill the current ABCI 1.0 requirements
+can still fulfill?
+Is this set of weaker requirements still strong enough to guarantee consensus _termination_?
+
+> Let's rephrase this a bit at the level of the consensus algorithm. Classically, Tendermint consensus 
+> has been analyzed using a function `valid(v, bc(h-1))` that ensured that correct processes agree on
+> the evaluation of this function because before starting height `h`, there was agreement on height 
+> `h - 1`. We want to relax that by using a function `valid(v, bc(h-1), x_p)` where `x_p` is a local 
+> variable at process `p`. As `x_p` may be different from `x_q` for two processes `p` and `q`, 
+> using `valid` in `PrepareProposal` 
+> and `ProcessProposal` may break Requirements 3 to 5.
+
+## Solution Proposed
+
+### Modified consensus _validity_
+The specification of consensus implicitly considers `valid(v)` as an immutable function,
+which is always supposed to provide the same result, when called at the same height
+for the same value `v`.
+This was the main reason for introducing the determinism requirements on `ProcessProposal`
+in the ABCI 1.0 specification.
+
+If we are to relax the determinism requirements on the application,
+we first need to weaken the `valid(v)` predicate and therefore consensus _validity_.
+In the proposed solution, correct processes MAY evaluate `valid(v)` to _true_ or _false_ at different times
+for the same value `v` during height `h`.
+
+Consensus _validity_ invariant is then modified as follows:
+
+- _weak validity_: predicate `valid(v)` has been evaluated to _true_ at least once by one correct process
+  for the decided block `v`.
+
+### Eventual requirements
+
+We now relax the relevant ABCI 1.0 requirements in the following way.
+
+* Requirement 3b [`PrepareProposal`, `ProcessProposal`, eventual coherence]:
+  There exists a round $r_s \ge 0$ of height $h$ such that,
+  for any two correct processes $p$ and $q$ and any round $r_p \geq r_s$,
+  if $q$'s CometBFT calls `RequestProcessProposal` on $u_p$,
```

Comment:
To be clear here: $u_p$ is produced at which round?

The definition of $u_p$ considers that it is produced at round $r_p$.

Reply:
Again, if we use a "local coherence" property, we could rephrase Requirement 5 with the inclusion of this stabilization round.

Reply:
Good points. Check my answer to @josef-widder [above](https://github.com/cometbft/knowledge-base/pull/14/files#r1317022025). I'm gonna try to get rid of rounds